lib/http_aws_sigv4.c

```diff
@@ -66,6 +68,159 @@ static void sha256_to_hex(char *dst, unsigned char *sha, size_t dst_l)
   }
 }
 
+
+
+static CURLcode make_headers(struct Curl_easy *data,
+                             const char *hostname,
+                             const char *timestamp,
+                             const char *provider1_low,
+                             const char *provider1_mid,
+                             char **canonical_headers,
+                             char **signed_headers)
+{
+  char *date_hdr_key =  curl_maprintf("X-%s-Date", provider1_mid);
+  char *date_full_hdr = curl_maprintf("x-%s-date:%s",
+                                      provider1_low, timestamp);
+#ifdef DEBUGBUILD
+  char *force_host = getenv("CURL_FORCEHOST");
+#endif
+  struct curl_slist *head = NULL;
+  struct curl_slist *tmp_head = NULL;
+  int ret = CURLE_OUT_OF_MEMORY;
+  struct curl_slist *l;
+  int again = 1;
+
+  if(!date_hdr_key || !date_full_hdr)
+    goto fail;
+
+  if(Curl_checkheaders(data, "Host")) {
+    head = NULL;
+  }
+#ifdef DEBUGBUILD
+  else if(force_host) {
+    head = curl_slist_append(NULL, "host:127.0.0.1");
+    if(!head)
+      goto fail;
+  }
+#endif
+  else {
+    char *full_host = data->state.aptr.host ? data->state.aptr.host :
+      curl_maprintf("host:%s", hostname);
+
+    if(!full_host)
+      goto fail;
+    head = curl_slist_append(NULL, full_host);
+    if(full_host != data->state.aptr.host)
+      free(full_host);
+    if(!head)
+      goto fail;
+  }
+
+
+  if(!date_hdr_key || !date_full_hdr)
+    goto fail;
+
+  for(l = data->set.headers; l; l = l->next) {
+    tmp_head = curl_slist_append(head, l->data);
+    if(!tmp_head)
+      goto fail;
+    head = tmp_head;
+  }
+
+  /* remove whitespace, and lowercase each headers */
+  for(l = head; l; l = l->next) {
+    char *tmp = l->data, *tmp2;
+
+    Curl_strntolower(l->data, l->data, strlen(l->data));
+    tmp = strchr(tmp, ':');
+    if(!tmp)
+      continue;
+    ++tmp;
+    tmp2 = tmp;
+    while(*tmp2 == ' ' || *tmp2 == '\t') {
+      ++tmp2;
+    }
+    if(tmp != tmp2)
+      memmove(tmp, tmp2, strlen(tmp2) + 1);
+
+    while(*tmp) {
+      tmp2 = tmp;
+      while(*tmp2 == ' ' || *tmp2 == '\t') {
+        ++tmp2;
+      }
+
+      if(tmp != tmp2) {
+        if(*tmp2 == 0)
+          /* remove spaces at the end */
+          *tmp = 0;
+        else {
+          *tmp++ = ' '; /* if tab convert to space */
+          if (*tmp == ' ' || *tmp == '\t')
+            memmove(tmp, tmp2, strlen(tmp2));
+        }
+        tmp = tmp2;
+      }
+      else {
+        ++tmp;
+      }
+    }
+
+  }
+
+  if(!Curl_checkheaders(data, date_hdr_key)) {
+    tmp_head = curl_slist_append(head, date_full_hdr);
+    if(!tmp_head)
+      goto fail;
+    tmp_head = head;
+  }
```

Comment:
If `X-Amz-Date` header is missed from the request headers, but present in CanonicalHeaders, how could the remote endpoint validate the signature?

Reply:
well, I have this code:
```
  if(!Curl_checkheaders(data, date_hdr_key)) {
    tmp_head = curl_slist_append(head, date_full_hdr);
    if(!tmp_head)
      goto fail;
    tmp_head = head;
  }
```
As, if I remember correctly AWS say you need to have X-Amz-Date, I force the header if not present.

Reply:
Sure, but first of all, headers in the request must be consistent with the headers in CanonicalHeaders. Here, you force the `X-Amz-Date` header only in CanonicalHeaders.

Reply:
yes your right, this code doesn't add it to the request header. (I mix up my code a little)
but the date is add to the request header by doing `data->state.aptr.userpwd = auth_headers` at the end of Curl_output_aws_sigv4.
as auth_headers contain `X-%s-Date`

Reply:
You are right: `auth_headers` has `X-%s-Date`, but I don't think it is fully correct. Here we consider two cases: 1) user has provided `X-%s-Date` headers 2) user has not provided the header. In the former case we will end up with the request consisted of two `X-%s-Date` headers (one from the user and one from `auth_headers`) with (most probably) different values, that highly likely will lead to the signature verification failure at the server side.

What do you think about moving `X-%s-Date` from `auth_headers` to `make_headers`?

Reply:
I think your right, also from my understand of 
```
For HTTP/1.1 requests, the host header must be included as a signed header. For HTTP/2 requests that include the :authority header instead of the host header, you must include the :authority header as a signed header. If you include a date or x-amz-date header, you must also include that header in the list of signed headers. 
```
if `date` is provide, I must not include `x-%s-date`, which is another corner case I don't handle yet.

I'm not sure I can move `auth_headers` into `make_headers` as it contain the whole signature, but I will not add `x-%s-date` all the time.

Reply:
Section "Example: PUT Object" of the spec contains the request with both `Date` and `X-Amz-Date` headers at the same time.